src/coprocessor_v2/raw_storage_impl.rs

```diff
@@ -0,0 +1,320 @@
+        Ok(())
+    }
+
+    async fn batch_put(&self, kv_pairs: Vec<KvPair>) -> StorageResult<()> {
+        let ctx = self.context.clone();
+        let cf = engine_traits::CF_DEFAULT.to_string();
+        let ttl = 0; // unlimited
+        let (cb, f) = paired_future_callback();
+
+        let res = self.storage.raw_batch_put(ctx, cf, kv_pairs, ttl, cb);
+
+        match res {
+            Err(e) => Err(e),
+            Ok(_) => f.await.map_err(StorageErrorShim::from)?,
+        }
+        .map_err(StorageErrorShim::from)?;
+        Ok(())
+    }
+
+    async fn delete(&self, key: Key) -> StorageResult<()> {
+        let ctx = self.context.clone();
+        let cf = engine_traits::CF_DEFAULT.to_string();
+        let (cb, f) = paired_future_callback();
+
+        let res = self.storage.raw_delete(ctx, cf, key, cb);
+
+        match res {
+            Err(e) => Err(e),
+            Ok(_) => f.await.map_err(StorageErrorShim::from)?,
+        }
+        .map_err(StorageErrorShim::from)?;
+        Ok(())
+    }
+
+    async fn batch_delete(&self, keys: Vec<Key>) -> StorageResult<()> {
+        let ctx = self.context.clone();
+        let cf = engine_traits::CF_DEFAULT.to_string();
+        let (cb, f) = paired_future_callback();
+
+        let res = self.storage.raw_batch_delete(ctx, cf, keys, cb);
+
+        match res {
+            Err(e) => Err(e),
+            Ok(_) => f.await.map_err(StorageErrorShim::from)?,
+        }
+        .map_err(StorageErrorShim::from)?;
+        Ok(())
+    }
+
+    async fn delete_range(&self, key_range: Range<Key>) -> StorageResult<()> {
+        let ctx = self.context.clone();
+        let cf = engine_traits::CF_DEFAULT.to_string();
+
+        let (cb, f) = paired_future_callback();
+
+        let res = self
+            .storage
+            .raw_delete_range(ctx, cf, key_range.start, key_range.end, cb);
+
+        match res {
+            Err(e) => Err(e),
+            Ok(_) => f.await.map_err(StorageErrorShim::from)?,
+        }
+        .map_err(StorageErrorShim::from)?;
+        Ok(())
+    }
+}
+
+/// Helper struct for converting between [`storage::errors::Error`] and
+/// [`coprocessor_plugin_api::StorageError`].
+struct StorageErrorShim(StorageError);
+
+impl From<StorageErrorShim> for StorageError {
+    fn from(err_shim: StorageErrorShim) -> Self {
+        err_shim.0
+    }
+}
+
+impl From<storage::errors::Error> for StorageErrorShim {
+    fn from(error: storage::errors::Error) -> Self {
+        let inner = match *error.0 {
+            // Key not in region
+            storage::errors::ErrorInner::Engine(EngineError(box EngineErrorInner::Request(
+                ref req_err,
+            ))) if req_err.has_key_not_in_region() => {
+                let key_err = req_err.get_key_not_in_region();
+                StorageError::KeyNotInRegion {
+                    key: key_err.get_key().to_owned(),
+                    region: todo!(), // TODO: how to construct region here? We only have region_id
```

Comment:
I'm not sure how we can construct a `Region` struct here because `StorageError::KeyNotInRegion` only has the `region_id`. How can we resolve a `region_id` to more information?